packages/amplify-graphql-relational-transformer/src/has-one/has-one-directive-ddb-references-transformer.ts

```diff
@@ -0,0 +1,47 @@
+import { DDB_DB_TYPE } from '@aws-amplify/graphql-transformer-core';
+import {
+  TransformerContextProvider,
+  TransformerPrepareStepContextProvider,
+  TransformerTransformSchemaStepContextProvider,
+} from '@aws-amplify/graphql-transformer-interfaces';
+import { DataSourceBasedDirectiveTransformer } from '../data-source-based-directive-transformer';
+import { DDBRelationalReferencesResolverGenerator } from '../resolver/ddb-references-generator';
+import { setFieldMappingResolverReference, updateTableForReferencesConnection } from '../resolvers';
+import { HasOneDirectiveConfiguration } from '../types';
+import { ensureReferencesArray, getReferencesNodes, registerHasOneForeignKeyMappings, validateParentReferencesFields } from '../utils';
+
+/**
+ * HasOneDirectiveDDBFieldsTransformer executes transformations based on `@hasOne(references: [String!])` configurations
+ * and surrounding TransformerContextProviders for DynamoDB data sources.
+ *
+ * This should not be used for `@hasOne(fields: [String!])` definitions.
+ */
+export class HasOneDirectiveDDBReferencesTransformer implements DataSourceBasedDirectiveTransformer<HasOneDirectiveConfiguration> {
+  dbType = DDB_DB_TYPE;
+
+  prepare = (context: TransformerPrepareStepContextProvider, config: HasOneDirectiveConfiguration): void => {
+    const modelName = config.object.name.value;
+    setFieldMappingResolverReference(context, config.relatedType?.name?.value, modelName, config.field.name.value, true);
+    registerHasOneForeignKeyMappings({
+      transformParameters: context.transformParameters,
+      resourceHelper: context.resourceHelper,
+      thisTypeName: modelName,
+      thisFieldName: config.field.name.value,
+      relatedType: config.relatedType,
+    });
+  };
+
+  transformSchema = (context: TransformerTransformSchemaStepContextProvider, config: HasOneDirectiveConfiguration): void => {
+    validateParentReferencesFields(config, context as TransformerContextProvider);
```

Comment:
why do we have validation steps here instead of in the `validate` method/step?

Reply:
Great question. This was already being called in the `transformSchema` method. [[reference](https://github.com/aws-amplify/amplify-category-api/blob/aa815ff1f6c54a8983c5250d58cfed4e7931aacd/packages/amplify-graphql-relational-transformer/src/graphql-has-one-transformer.ts#L192)] This isn't the only case of validation being done outside of `validate`. I'm tracking this work in the [feature branch PR](https://github.com/aws-amplify/amplify-category-api/pull/2370):
>  Move all necessary validation to `validate` methods in relational directive transformers.